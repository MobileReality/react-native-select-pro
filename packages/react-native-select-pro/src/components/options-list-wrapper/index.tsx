import React, { ComponentProps, ReactNode, useEffect, useRef } from 'react';
import { Animated, StyleProp, View, ViewStyle } from 'react-native';

import type { OptionalToRequired } from '../../helpers/types/OptionalToRequired';
import type { OptionsList } from '../options-list';

type FromOptionListProps = Pick<
    ComponentProps<typeof OptionsList>,
    'animated' | 'animationDuration' | 'isOpened'
>;

type WrapperStyles = {
    wrapperStyles: StyleProp<ViewStyle>;
};

type OptionsListWrapperProps = OptionalToRequired<FromOptionListProps> & {
    children: ReactNode;
} & WrapperStyles;

export const OptionsListWrapper = ({
    children,
    animated,
    animationDuration,
    isOpened,
    wrapperStyles,
}: OptionsListWrapperProps) => {
    const fadeAnimation = useRef(new Animated.Value(0)).current;

    useEffect(() => {
        if (animated) {
            Animated.timing(fadeAnimation, {
                toValue: isOpened ? 1 : 0,
                duration: animationDuration,
                useNativeDriver: true,
            }).start();
        }
    }, [fadeAnimation, isOpened]);

    return animated ? (
        <Animated.View
            pointerEvents={isOpened ? 'auto' : 'none'}
            style={[wrapperStyles, { opacity: fadeAnimation }]}
        >
            {children}
        </Animated.View>
    ) : isOpened ? (
        <View style={wrapperStyles}>{children}</View>
    ) : null;
};
