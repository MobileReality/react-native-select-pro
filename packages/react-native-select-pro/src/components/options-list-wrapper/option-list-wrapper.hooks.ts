import { useOptionsListContext } from '../../context';
import { useAnimation } from '../../hooks';

export const useOptionListWrapper = () => {
    const {
        animation,
        isOpened,
        styles,
        aboveSelectControl,
        openedPosition: { top, left, width },
    } = useOptionsListContext();

    const { optionsList: optionsListCustomStyles } = styles ?? {};

    const fadeAnimation = useAnimation({ isOpened, animation });

    return {
        animation,
        isOpened,
        fadeAnimation,
        optionsListCustomStyles,
        aboveSelectControl,
        top,
        left,
        width,
    };
};
