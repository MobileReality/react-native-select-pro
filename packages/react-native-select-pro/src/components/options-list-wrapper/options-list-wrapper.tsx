import React, { forwardRef } from 'react';
import type { ViewStyle } from 'react-native';
import { Animated, StyleSheet, View } from 'react-native';

import { BORDER_WIDTH, COLORS, OPTIONS_LIST_HEIGHT, SHAPE } from '../../constants';

import { useOptionListWrapper } from './option-list-wrapper.hooks';
import type { OptionsListWrapperProps } from './options-list-wrapper.types';

export const OptionsListWrapper = forwardRef<View, OptionsListWrapperProps>(({ children }, ref) => {
    const {
        isOpened,
        fadeAnimation,
        optionsListCustomStyles,
        aboveSelectControl,
        top,
        left,
        width,
    } = useOptionListWrapper();

    return fadeAnimation ? (
        <Animated.View
            ref={ref}
            pointerEvents={isOpened ? 'auto' : 'none'}
            style={[
                styles.optionsList,
                aboveSelectControl ? styles.overflown : styles.notOverflown,
                optionsListCustomStyles,
                { top, left, width },
                { opacity: fadeAnimation },
            ]}
        >
            {children}
        </Animated.View>
    ) : isOpened ? (
        <View
            ref={ref}
            style={[
                styles.optionsList,
                aboveSelectControl ? styles.overflown : styles.notOverflown,
                optionsListCustomStyles,
                { top, left, width },
            ]}
        >
            {children}
        </View>
    ) : null;
});

type Styles = {
    optionsList: ViewStyle;
    notOverflown: ViewStyle;
    overflown: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    optionsList: {
        flex: 1,
        position: 'absolute',
        zIndex: 1,
        backgroundColor: COLORS.WHITE,
        borderWidth: BORDER_WIDTH,
        maxHeight: OPTIONS_LIST_HEIGHT,
        elevation: 5,
    },
    notOverflown: {
        borderTopWidth: 0,
        borderBottomRightRadius: SHAPE,
        borderBottomLeftRadius: SHAPE,
    },
    overflown: {
        borderBottomWidth: 0,
        borderTopRightRadius: SHAPE,
        borderTopLeftRadius: SHAPE,
    },
});

OptionsListWrapper.displayName = 'OptionsListWrapper';
