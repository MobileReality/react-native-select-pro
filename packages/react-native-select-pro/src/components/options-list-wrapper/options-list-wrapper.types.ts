import type { ReactNode } from 'react';

export type OptionsListWrapperProps = {
    children: ReactNode;
};
