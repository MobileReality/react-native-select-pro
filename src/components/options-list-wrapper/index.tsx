import React, { ComponentProps, ReactElement, useRef } from 'react';
import { Animated, StyleProp, View, ViewStyle } from 'react-native';

import { ANIMATION_DURATION } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers';
import type { Select } from '../../index';
import type { OptionsList } from '../options-list';

type FromSelectProps = Pick<ComponentProps<typeof Select>, 'isAnimated'>;
type FromOptionListProps = Pick<ComponentProps<typeof OptionsList>, 'isOpened'>;

type WrapperStyles = {
    wrapperStyles: StyleProp<ViewStyle>;
};

type OptionsListWrapperProps = OptionalToRequired<FromSelectProps & FromOptionListProps> & {
    children: ReactElement;
} & WrapperStyles;

export const OptionsListWrapper = ({
    children,
    isAnimated,
    isOpened,
    wrapperStyles,
}: OptionsListWrapperProps) => {
    const fadeAnimation = useRef(new Animated.Value(0)).current;

    React.useEffect(() => {
        if (isAnimated) {
            Animated.timing(fadeAnimation, {
                toValue: isOpened ? 1 : 0,
                duration: ANIMATION_DURATION,
                useNativeDriver: true,
            }).start();
        }
    }, [fadeAnimation, isOpened]);

    return isAnimated ? (
        <Animated.View
            pointerEvents={isOpened ? 'auto' : 'none'}
            style={[wrapperStyles, { opacity: fadeAnimation }]}>
            {children}
        </Animated.View>
    ) : isOpened ? (
        <View style={wrapperStyles}>{children}</View>
    ) : null;
};
