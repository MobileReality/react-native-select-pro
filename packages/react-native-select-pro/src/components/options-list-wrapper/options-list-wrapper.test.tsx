import React from 'react';
import { Text, View } from 'react-native';
import { render } from '@testing-library/react-native';

import { OptionsListWrapper } from './options-list-wrapper';

describe('OptionsListWrapper', () => {
    it('should generate OptionsListWrapper snapshot', () => {
        const optionsListWrapper = render(
            <OptionsListWrapper animation={false} isOpened={false} wrapperStyles={{}}>
                <View>
                    <Text>Test</Text>
                </View>
            </OptionsListWrapper>,
        );
        expect(optionsListWrapper).toMatchSnapshot();
    });
});
