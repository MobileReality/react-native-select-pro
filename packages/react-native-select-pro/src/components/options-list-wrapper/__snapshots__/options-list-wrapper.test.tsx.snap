// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsListWrapper should generate OptionsListWrapper snapshot 1`] = `null`;
